crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -387,6 +383,87 @@ impl QueryBuilder {
+                    events::dsl::tx_sequence_number.lt(cursor_val.0).or(
+                        events::dsl::tx_sequence_number
+                            .eq(cursor_val.0)
+                            .and(events::dsl::event_sequence_number.lt(cursor_val.1)),
+                    ),
+                );
+            } else {
+                query = query.filter(
+                    events::dsl::tx_sequence_number.gt(cursor_val.0).or(
+                        events::dsl::tx_sequence_number
+                            .eq(cursor_val.0)
+                            .and(events::dsl::event_sequence_number.gt(cursor_val.1)),
+                    ),
+                );
+            }
+        }
+        if descending_order {
+            query = query.order((
+                events::dsl::tx_sequence_number.desc(),
+                events::dsl::event_sequence_number.desc(),
+            ));
+        } else {
+            query = query.order((
+                events::dsl::tx_sequence_number.asc(),
+                events::dsl::event_sequence_number.asc(),
+            ));
+        }
+        query = query.limit(limit + 1);
+
+        if let Some(filter) = filter {
+            if let Some(sender) = filter.sender {
+                // Construct a subquery to filter on senders - this is because we do not have an index on the senders column.
+                let subquery = tx_senders::dsl::tx_senders
+                    .filter(tx_senders::dsl::sender.eq(sender.into_vec()))
+                    .select(tx_senders::dsl::tx_sequence_number);
+
+                query = query.filter(events::dsl::tx_sequence_number.eq_any(subquery));
+            }
+
+            if let Some(digest) = filter.transaction_digest {
+                let tx_digest = Digest::from_str(&digest)?.into_vec();
+                let subquery = transactions::dsl::transactions
+                    .filter(transactions::dsl::transaction_digest.eq(tx_digest))
+                    .select(transactions::dsl::tx_sequence_number);
+
+                query = query.filter(events::dsl::tx_sequence_number.eq_any(subquery));
+            }
+
+            if let Some(p) = filter.emitting_package {
+                query = query.filter(events::dsl::package.eq(p.into_vec()));
+
+                if let Some(m) = filter.emitting_module {
+                    query = query.filter(events::dsl::module.eq(m));
+                }
+            }
+
+            if let Some(p) = filter.event_package {
+                if let Some(m) = filter.event_module {
+                    if let Some(t) = filter.event_type {
+                        query = query
+                            .filter(events::dsl::event_type.like(format!("{}::{}::{}%", p, m, t)));
+                    }
+                    query = query.filter(events::dsl::event_type.like(format!("{}::{}::%", p, m)));
+                }
+                query = query.filter(events::dsl::event_type.like(format!("{}::%", p)));
+            }
+        }
```

Comment:
nit: this `if` statement could be merged with the above, right?

Reply:
Probably just misunderstood, but as in merging the logic for emittingPackage and eventPackage together? i thought eventtype and emitting type filters were independent filters

Reply:
Ah no, this is on me -- I didn't spot the difference between `emitting`  and `event` here!  Could you stick a comment on these nested ifs to call that out with a little explanation of what each one is?